x/rewards/keeper/iprpc.go

```diff
@@ -0,0 +1,170 @@
+	nextMonthIprpcReward, found := k.PopIprpcReward(ctx, false)
+	nextMonthId := k.GetIprpcRewardsCurrent(ctx)
+	if !found {
+		nextMonthIprpcReward = types.IprpcReward{Id: nextMonthId, SpecFunds: iprpcReward.SpecFunds}
+	} else {
+		nextMonthIprpcReward.SpecFunds = k.transferSpecFundsToNextMonth(iprpcReward.SpecFunds, nextMonthIprpcReward.SpecFunds)
+	}
+	k.SetIprpcReward(ctx, nextMonthIprpcReward)
+	details := map[string]string{
+		"transferred_funds":        iprpcReward.String(),
+		"next_month_updated_funds": nextMonthIprpcReward.String(),
+	}
+	utils.LogLavaEvent(ctx, k.Logger(ctx), types.TransferIprpcRewardToNextMonth, details,
+		"No provider serviced an IPRPC eligible subscription, transferring current month IPRPC funds to next month")
+}
+
+// countIprpcCu updates the specCuMap which keeps records of spec->SpecCuType (which holds the IPRPC CU per provider)
+func (k Keeper) countIprpcCu(specCuMap map[string]types.SpecCuType, iprpcCu uint64, spec string, provider string) {
+	if iprpcCu != 0 {
+		specCu, ok := specCuMap[spec]
+		if !ok {
+			specCuMap[spec] = types.SpecCuType{
+				ProvidersCu: map[string]uint64{provider: iprpcCu},
+				TotalCu:     iprpcCu,
+			}
+		} else {
+			_, ok := specCu.ProvidersCu[provider]
+			if !ok {
+				specCu.ProvidersCu[provider] = iprpcCu
+			} else {
+				specCu.ProvidersCu[provider] += iprpcCu
+			}
+			specCu.TotalCu += iprpcCu
+			specCuMap[spec] = specCu
+		}
+	}
+}
+
+// AddSpecFunds adds funds for a specific spec for <duration> of months.
+// This function is used by the fund-iprpc TX.
+func (k Keeper) addSpecFunds(ctx sdk.Context, spec string, fund sdk.Coins, duration uint64) {
+	startID := k.GetIprpcRewardsCurrent(ctx) + 1 // fund IPRPC only from the next month for <duration> months
+	for i := startID; i < startID+duration; i++ {
+		iprpcReward, found := k.GetIprpcReward(ctx, i)
+		if found {
+			// found IPRPC reward, find if spec exists
+			specIndex := -1
+			for i := 0; i < len(iprpcReward.SpecFunds); i++ {
+				if iprpcReward.SpecFunds[i].Spec == spec {
+					specIndex = i
+				}
+			}
+			// update spec funds
+			if specIndex >= 0 {
+				iprpcReward.SpecFunds[specIndex].Fund = iprpcReward.SpecFunds[specIndex].Fund.Add(fund...)
+			} else {
+				iprpcReward.SpecFunds = append(iprpcReward.SpecFunds, types.Specfund{Spec: spec, Fund: fund})
+			}
+		} else {
+			// did not find IPRPC reward -> create a new one
+			iprpcReward.Id = i
+			iprpcReward.SpecFunds = []types.Specfund{{Spec: spec, Fund: fund}}
+		}
+		k.SetIprpcReward(ctx, iprpcReward)
+	}
+}
+
+// transferSpecFundsToNextMonth transfer the specFunds to the next month's IPRPC funds
+// this function is used when there are no providers that should get the monthly IPRPC reward,
+// so the reward transfers to the next month
+func (k Keeper) transferSpecFundsToNextMonth(specFunds []types.Specfund, nextMonthSpecFunds []types.Specfund) []types.Specfund {
```

Comment:
do it simpler, 2 for loops
we dont have that many specs

Reply:
done. see https://github.com/lavanet/lava/commit/e1da79c790caf2ae562891f9ef3b974b1c009986